models/src/main/scala/coop/rchain/models/rholangn/parmanager/Manager.scala

```diff
@@ -59,16 +61,25 @@ object Manager {
   def combinePars(p1: ParN, p2: ParN): ParN = flattedPProc(Seq(p1, p2))
 
   /** MetaData */
-  def rhoHashFn(p: RhoTypeN): Array[Byte]      = RhoHash.rhoHashFn(p)
+  def rhoHashFn(p: RhoTypeN): Eval[Array[Byte]] = {
+    val write = (out: CodedOutputStream) => RhoHash.serializeForHash(p, ProtoPrimitiveWriter(out))
+    // Return data padded with zero from left if its size less then the limit ot hash of the data
+    def padOrHash(data: Array[Byte], limit: Int, hash: Array[Byte] => Array[Byte]): Array[Byte] = {
+      val pad = limit - data.length
+      if (pad <= 0) hash(data) else Array.concat(data, Array.fill(pad)(0.toByte))
+    }
+    // 4096 is the preallocated size of ByteArrayOutputStream, might be enough for most cases
+    ProtoCodec.encode(4096, write).map { padOrHash(_, hashSize, Blake2Hash.hash) }
```

Comment:
The size of the byte stream will be the serialized primitive types or collection of hashes from nested objects. So maybe better that we go lower with initial size like 1024 or even half of that. The size is growing by doubling so for bigger arrays it should still be minimal overhead.